lib/utils/authentication.js

```diff
@@ -44,23 +45,114 @@ module.exports.register = function(server, options, next) {
      * Even though this is a synchronous function, it returns a promise
      * This will allow for an easy transition to async key generators
      * @param {string} username - unencrypted username
-     * @return {Bluebird} resolves to a credentials object
+     * @return {Promise} resolves to a credentials object
      *   where credentials is an object with the following properties:
-     *   `id, key, algorithm, issueTime, expireTime`
+     *   `id, key, algorithm, issueTime, expireTime, studyRoles`
      */
     authUtils.generateHawkCredentials = function(username) {
-        const credentials = {
-            username: username,
-            id: uuid.v4(),
-            key: uuid.v4(),
-            algorithm: config.algorithm,
-            issueTime: +new Date(),
-            expireTime: +new Date() + config.hawkKeyLifespan
-        };
-        return Bluebird.resolve(credentials);
+
+        return refreshCurrentStudyRoles(username)
+            .then((roleList) => {
+                const credentials = {
+                    username: username,
+                    id: uuid.v4(),
+                    key: uuid.v4(),
+                    algorithm: config.algorithm,
+                    issueTime: + new Date(),
+                    expireTime: + new Date() + config.hawkKeyLifespan,
+                    studyRoles: roleList
+                };
+
+                return credentials;
+            });
     };
 
     /**
+     * Retrieves the current roles for a user, revokes old roles, and
+     * adds the new roles
+     * @param {string} username - the username to retrieve roles
+     * @return {object} roles - the current roles
+     */
+    function refreshCurrentStudyRoles(username) {
+        const encryptedUsername = authUtils.encryptUsername(username);
+
+        const UserRole = server.plugins.bookshelf.model('UserStudyRole');
+    
+        return new UserRole()
+            .where({username: encryptedUsername})
+            .fetchAll({withRelated: 'role'})
+            .then((roleCollection) => {
```

Comment:
ok, I may have just figured out a cleaner way to do this...

You are trying to perform the following steps:
1. fetch `roleCollection`
1. revokeRoles from username
1. add new roles to username
1. fetch list of `studyRoles` for user

The steps are each async, and must occur in series to prevent accidentally revoking the newly set roles.

Option 1: _Create a resetStudyRoles function that performs steps 2,3
```
const resetStudyRoles = (roleCollection) => {
  return revokeRoles(username)
    .then(_.partial(addRolesFromCollection, username, roleCollection));
  };

 const getStudyRoles = () => {
   return relations.studyAsync('Describe what %s can do', username);
 };

 return new UserRole()
   .where({username: encryptedUsername})
   .fetchAll({withRelated: 'role'})
   .then(resetStudyRoles)
   .then(getStudyRoles)
   .catch(...);
```

Option 2: (specific to this route) If the order of steps 1 and 2 can be reversed...
```
 const getStudyRoles = () => {
   return relations.studyAsync('Describe what %s can do', username);
 };

 return new UserRole()
   .where({username: encryptedUsername})
   .fetchAll({withRelated: 'role'})
   .then(_.partial(addRolesFromCollection, username))
   .then(_.partial(revokeRoles, username))
   .then(getStudyRoles)
   .catch(...);
```